add hessian kernel for kappa and mu

wjlei1990 opened the pull request.

danielpeter committed:
updates interpolation (avoiding sqrt() for radius)

danielpeter committed:
fixes attenuation freq variable for simultaneous runs

danielpeter committed:
fixes gravity indexing (for earth)

danielpeter committed:
fixes indexing for fluid properties in gravity

danielpeter committed:
adds optional mute input

danielpeter committed:
fixing declarations

danielpeter committed:
updates HDUR_MOVIE calculation

danielpeter committed:
updates shifted velocity output

wjlei1990 committed:
add hessian kernel for kappa and mu

1sbeller committed:
Update get_model.F90

Add absolute value of argument to avoid square root of slightly negative values.

1sbeller committed:
Update get_model.F90

danielpeter committed:
Merge pull request #691 from 1sbeller/patch-1

Solve mesher crash on power9 architectures

danielpeter committed:
adding optional input to use moving average or absolute scale

danielpeter committed:
correcting Gc,Gs prime kernel w/ multiplication of mu

1sbeller committed:
Check averaged isotropic shear wavespeed computations for all models.

1sbeller committed:
remove non ascii characters and some useless vs checks

1sbeller committed:
Manage STRINGIFY macro for PGI compiler

So that ADIOS1 and the mesher runs when using PORTLAND compilers.

1sbeller committed:
Compiles model_vpremoon before  get_model_parameters to avoid undefined reference with PGI

Now mesher and solver both works on PORTLAND compiler.

danielpeter committed:
Merge pull request #693 from 1sbeller/patch-1

Check averaged isotropic shear wavespeed computations for all models.

wjlei1990 committed:
add hessian kernels for rho, kappa and mu

wjlei1990 committed:
add write functions for hessians of rho, kappa and mu

danielpeter commented on src/specfem3D/compute_kernels.F90:
```diff
@@ -860,13 +860,25 @@ subroutine compute_kernels_Hessian()
 #endif
+  integer :: i,j,k
+
+  interface
```
so far, we don't have many such interface definitions in the code. i would say unless there is really a need for it (like for example a function definition returning an array), you can skip this part.

wjlei1990 replied:
Hi Daniel, I think I have to define the interface explicitly, otherwise if I delete the interface, the compile will faild:
```
Error: Procedure 'compute_gradient_crust_mantle' at (1) with assumed-shape dummy argument 
'field' must have an explicit interface
```
Since the input argument `field` is defined as
```
real(kind=CUSTOM_REAL), dimension(:, :), intent(in) :: field
```

The other way around would be define the function in a module.

danielpeter replied:
right, i missed that you had left field without dimensions. 

we usually assign field sizes like `..,dimension(NDIM,nglob)` and either pass nglob as argument or use the ones saved in the mesher file `values_from_mesher.h` like `NGLOB_CRUST_MANTLE`. that way, we can omit the interface definitions.

wjlei1990 replied:
Thanks. Fixed with explicitly specifying the array dimension.

danielpeter commented on src/specfem3D/compute_kernels.F90:
```diff
@@ -889,13 +901,63 @@ subroutine compute_kernels_Hessian()
 
         iglob = ibool_crust_mantle(INDEX_IJK,ispec)
 
+#ifdef FORCE_VECTORIZATION
```
above you will find the DO_LOOP_IJK macro. the macro only makes sense if you don't really need i,j,k explicitly but can loop over an indexing like (ijk,1,1,..). in those cases, the compiler can do vectorization and the speed performance becomes a bit better. so, here the kernel calculations are not really performance critical and if you need i,j,k indices, just write out the do-loops do i=..; do j= ..; do k=.. and omit the INDEX_IJK notation completely.

danielpeter commented on src/specfem3D/compute_kernels.F90:
```diff
@@ -889,13 +901,63 @@ subroutine compute_kernels_Hessian()
 
+        call calculate_spatial_deriv_crust_mantle(veloc_crust_mantle, veloc_deriv, i, j, k, ispec)
+
+        if (.NOT. USE_SOURCE_RECEIVER_HESSIAN) then
```
another detail, if you have the DO_LOOP_IJK section, don't add an if-statement into the inner-most loop. this will most likely break the vectorization of the code. thus instead, put the if-statement outside:
```
if (USE_SOURCE..) then
  DO_LOOP_IJK
   ..
  ENDDO_LOOP_IJK
else
  DO_LOOP_IJK
  ..
```
etcetera.

also, why not use a positive if-statement: if (USE_SOURCE_RECEIVER..) instead of if (.NOT. ..)? this makes it easier to read the code. 

only exception: if you already know that the flag mostly is negative, then start with the more likelier if-statement. this will evaluate first and the code becomes faster.

danielpeter commented on src/specfem3D/compute_kernels.F90:
```diff
@@ -889,13 +901,63 @@ subroutine compute_kernels_Hessian()
+                  + (veloc_deriv(1, 3) + veloc_deriv(3, 1)) * (veloc_deriv(1, 3) + veloc_deriv(3, 1))
+        else
+          ! use source-source hessian kernels
+          call calculate_spatial_deriv_crust_mantle(b_veloc_crust_mantle, b_veloc_deriv, i, j, k, ispec)
```
maybe better to write a compute the gradient routine per one full element ispec, instead of per each gll point i,j,k. calling it per element allows to compute it before the heavy small do i=.. loops and reading in the xix_.. becomes more cache efficient.

danielpeter commented on src/specfem3D/compute_kernels.F90:
```diff
@@ -911,3 +973,112 @@ subroutine compute_kernels_Hessian()
+      tempz3l = tempz3l + field(3, iglob) * hprime_zz(k, l)
+    enddo
+
+    xixl = deriv_mapping_crust_mantle(1, i, j, k, ispec)
```
the deriv_mapping_** array is only available if (USE_DEVILLE_PRODUCTS_VAL) is true. so would need this if-statement, else you need to use the xix_** arrays.

danielpeter committed:
adding options to order mpi assembly and output seismograms of the adjoint wavefield (instead of reconstructed backward wavefield)

danielpeter committed:
updates LDDRK scheme (usable for both forward/kernel runs)

danielpeter committed:
updates gravity on gpus using pre-calculated arrays

danielpeter committed:
adds a compilation dependency

danielpeter committed:
fixes argument in debugging call

Commit:
Buildbot ran the code beautifier.

danielpeter committed:
updates references in REF_SEIS/ for example regional_s40rts/

wjlei1990 committed:
add function compute_gradient_crust_mantle to calculate gradient at element level

danielpeter committed:
updates combine_vol_data to work with double precision arrays

danielpeter committed:
adds example for higher-order time stepping (EXAMPLES/regional_Greece_small_LDDRK/)

wjlei1990 committed:
change dummary argument name

wjlei1990 committed:
remove interface by explicitly specifying argument array dimension

danielpeter committed:
updates mesh2vtu

wjlei1990 committed:
add GPU code to compute hessian kernels for rho, kappa and mu

wjlei1990 committed:
change float to realw in GPU code

wjlei1990 committed:
fix bug in reference code for computing hess kernels

wjlei1990 committed:
add BOAST gpu code for computing extra hessian kernels

wjlei1990 committed:
add opencl gpu code for computing extra hessian kernels

danielpeter committed:
updates output for shifted model values; adds check for gll model files for inner core; updates attenuation comments

danielpeter committed:
renames parameter in Par_file

danielpeter committed:
renames master/slave -> main/secondary (according to new linux naming scheme)

danielpeter committed:
code cleaning

danielpeter committed:
updates user outputs

danielpeter committed:
updates noise example

danielpeter committed:
adds travis example

danielpeter committed:
fixes not-equal comparisons in c files

danielpeter committed:
updates images in noise example

danielpeter committed:
updates doc

danielpeter committed:
updates travis

danielpeter committed:
updates model SGLOBE-rani

danielpeter committed:
adds test example for SGLOBE-rani

danielpeter committed:
makes sgloberani example regional

danielpeter committed:
reduces default timestep for sglobe crust

danielpeter committed:
cleans out ref files

danielpeter committed:
adds LDDRK fields to GPU (missing still full attenuation)

danielpeter committed:
adds attenuation to LDDRK for GPU simulations

danielpeter committed:
fixes opencl call

danielpeter committed:
adds github action workflow

danielpeter committed:
updates workflow

danielpeter committed:
more updates on array allocation (to avoid stack issues with newer gcc compilers version 10.x on MacOS)

danielpeter committed:
updates a comment

wjlei1990 committed:
Bug fix when calculating kappa and mu hessian kernels

Add code to copy the kerneles from GPU back to CPU

danielpeter committed:
updates rules

danielpeter committed:
avoiding mpi calls for modules used in xcreate_header_file (to be able to run/link without MPI)

Commit:
Buildbot ran the code beautifier.

icui committed:
Option to use monochromatic source for CMTSOLUTION

icui committed:
Merge remote-tracking branch 'upstream/devel' into devel

icui committed:
Updated Par_file in EXAMPLES

Commit:
Buildbot ran the code beautifier.

danielpeter committed:
updates time step size for sglobe crustal models

danielpeter committed:
updates Par_file

bangerth committed:
Fix references to a function name in the manual.

The name of the function documented in the manual is `model_s20rts`, but the actual name of the function in the code is `mantle_s20rts`. While there also clarify two sentences that talk about files but use the word subroutine.

icui committed:
Added an option to impose a sponge area

icui committed:
Updated Par_file for EXAMPLES

Commit:
Buildbot ran the code beautifier.

icui committed:
Fixed an uninitialized variable in xsmooth_sem

wjlei1990 committed:
change the math formula of hessian kernels for kappa and mu

slgogar committed:
Improving OpenCL host to device data transfer using clEnqueueWriteBufferRect

Commit:
Buildbot ran the code beautifier.

wjlei1990 committed:
Merge remote-tracking branch 'upstream/devel' into hessian
